### PR TITLE
Coordinator for training set, materialize and update redis for types

### PR DESCRIPTION
**Pull request by Sami1309**
Notes about testing this:

coordinator only works with EtcdLookup, not memory lookup for metadata
Needs consistent provider (no in memory provider)
etcd needs to be running in deployment

Currently it's a very hacky test, needs the metadata server to be running. Should have it so it spawns it's own.

TODO:
Test set attempts function (does it set the attempts correctly?

1. Abstract integration tests out to every relevant provider (extend to multiple tables/features/labels, error cases, other providers)
6. Get to 85% coverage

### CODE REVIEW

File: ff/serving/coordinator/coordinator.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	didDelete := response.Responses[0].GetResponseDeleteRange().Deleted
+	if didDelete != 1 {
+		return fmt.Errorf("Job Already deleted")
+	}
+	return nil
+}
+
+func (c *Coordinator) syncHandleJob(jobKey string, s *concurrency.Session) error {
+	mtx := concurrency.NewMutex(s, fmt.Sprintf("LOCK_%s", job_key))
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:237]: move Sprintf to its own function or at the very least a variable

Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:237]: also job_key -> jobKey

File: ff/serving/coordinator/coordinator.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Could not deserialize coordinator job: %v", err)
+	}
+	if job.Attempts > MAX_ATTEMPTS {
+		if err := c.SetResourceStatus(job.Resource, metadata.ResourceStatus.Failed); err != nil {
+			return fmt.Errorf("Could not set job status to failed: %v", err)
+		}
+		return nil
+	}
+	if err := c.incrementJobInETCD(mtx, job, jobKey); err != nil {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:258]: incrementJob

File: ff/serving/coordinator/coordinator.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+		return err
+	}
+	switch job.Resource.Type {
+	case metadata.TRAINING_SET_VARIANT:
+		if err := c.runTrainingSetJob(job.Resource); err != nil {
+			return fmt.Errorf("Training set job failed: %v", err)
+		}
+	}
+	if err := c.safeDelete(mtx, jobKey); err != nil {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:267]: deleteJob

File: ff/serving/coordinator/coordinator.go (hunk 1 above)
Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:237]: You could make this its own fucntion too, createJobLock

File: ff/serving/coordinator/coordinator.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+		return fmt.Errorf("Delete job transaction failed")
+	}
+	didDelete := response.Responses[0].GetResponseDeleteRange().Deleted
+	if didDelete != 1 {
+		return fmt.Errorf("Job Already deleted")
+	}
+	return nil
+}
+
+func (c *Coordinator) syncHandleJob(jobKey string, s *concurrency.Session) error {
```
[the remark this answers is not in the document]
Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:236]: executeJob*

File: ff/serving/coordinator/coordinator.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+		}
+	}()
+	value, err := c.safeGet(mtx, jobKey)
+	job := &CoordinatorJob{}
+	err := job.Deserialize(responseData)
+	if err != nil {
+		return nil, fmt.Errorf("Could not deserialize coordinator job: %v", err)
+	}
+	if job.Attempts > MAX_ATTEMPTS {
+		if err := c.SetResourceStatus(job.Resource, metadata.ResourceStatus.Failed); err != nil {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:253]: markJobFailed

File: ff/serving/coordinator/coordinator.go (hunk 2 above)
Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:258]: incrementJobAttempts

File: ff/serving/coordinator/coordinator.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	keyValueClient := clientv3.NewKV(c.ETCDClient)
+	txn := (keyValueClient).Txn(context.Background())
+	if _, err := txn.If(l.IsOwner()).Then(clientv3.OpPut(jobKey, serialized)).Commit(); err != nil {
+		return fmt.Errorf("Could not set iterated coordinator job. %v", err)
+	}
+	return nil
+}
+
+func (c *Coordinator) safeDelete(mtx *concurrency.Mutex, key string) error {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:219]: deleteJob

File: ff/serving/coordinator/coordinator.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+	keyValueClient := clientv3.NewKV(c.ETCDClient)
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:220]: why not just create this once when you create the coordinator and make it a field

File: ff/serving/coordinator/coordinator.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	txn = (keyValueClient).Txn(context.Background())
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:221]: no paratheses

Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:221]: :=?

File: ff/serving/coordinator/coordinator.go (hunk 7 above)
Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:220]: rename to kvClient? or kvInterface, even etcdKVInterface

File: ff/serving/coordinator/coordinator.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	response, err := txn.If(mtx.IsOwner()).Then(clientv3.OpDelete(jobKey)).Commit()
+	if err != nil {
+		return fmt.Errorf("Delete job transaction failed: %v", err)
+	}
+	if !response.Succeeded {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:226]: isOwner := response.Succeeded, add a comment

File: ff/serving/coordinator/coordinator.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+		return fmt.Errorf("Delete job transaction failed")
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:227]: Delete job transaction failed: coordinator does not own job

File: ff/serving/coordinator/coordinator.go [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	didDelete := response.Responses[0].GetResponseDeleteRange().Deleted
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:229]: add the != 1, also avoid magic numbers like [0], you could break this into 3 variables

Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:229]: .GetResponseDeleteRange().Deleted = numDeleted

File: ff/serving/coordinator/coordinator.go [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+func (c *Coordinator) incrementJobInETCD(mtx *concurrency.Mutex, job CoordinatorJob, jobKey string) error {
+	job.Attempts += 1
+	serializedJob, err := job.Serialize()
+	if err != nil {
+		return fmt.Errorf("Could not serialize coordinator job. %v", err)
+	}
+	keyValueClient := clientv3.NewKV(c.ETCDClient)
+	txn := (keyValueClient).Txn(context.Background())
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:212]: no parens

File: ff/serving/coordinator/coordinator.go [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+	if _, err := txn.If(l.IsOwner()).Then(clientv3.OpPut(jobKey, serialized)).Commit(); err != nil {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:213]: What is l

Comment (simba-git) [reply ff/serving/coordinator/coordinator.go:213]: take the L

File: ff/serving/coordinator/coordinator.go [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+	p, err := provider.Get(provider.Type(providerEntry.Type()), providerEntry.SerializedConfig())
+	if err != nil {
+		return err
+	}
+	store, err := p.AsOfflineStore()
+	if err != nil {
+		return err
+	}
+	providerResID := provider.ResourceID{Name: resID.Name, Variant: resID.Variant}
+	if _, err := store.GetTrainingSet(providerResID); err == nil {
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:146]: You can't delete a training set so this doesn't have to be atomic

File: ff/serving/coordinator/coordinator.go [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+	runner "github.com/featureform/serving/runner"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+)
+
+type Coordinator struct {
+	Metadata   *metadata.Client
+	Logger     *zap.SugaredLogger
+	ETCDClient *clientv3.Client
+	Spawner    JobSpawner
```
Comment (simba-git) [inline ff/serving/coordinator/coordinator.go:21]: Make JobSpawner an interface

**Review by simba-git — CHANGES_REQUESTED** [review]

File: ff/serving/metadata/client.go [hunk 16]
```diff
@@ -395,6 +404,7 @@ func (client *Client) CreateTrainingSetVariant(ctx context.Context, def Training
 		Description: def.Description,
 		Owner:       def.Owner,
 		Provider:    def.Provider,
+		Status:      string(CREATED),
```
Comment (Sami1309) [inline ff/serving/metadata/client.go:407]: labels wouldn't have a status, so should be NO_STATUS

File: ff/serving/metadata/metadata.go [hunk 17]
```diff
@@ -1103,6 +1215,11 @@ func (serv *MetadataServer) genericCreate(ctx context.Context, res Resource, ini
 	if err := serv.lookup.Set(id, res); err != nil {
 		return nil, err
 	}
+	if serv.needsJob(res) {
```
Comment (Sami1309) [inline ff/serving/metadata/metadata.go:1218]: should make this atomic so a resource isn't set without the job being set
resource interface should have a function HasJob, and it can have a SetJob so that the resource is set with the job atomically

File: ff/serving/metadata/proto/metadata.proto [hunk 18]
```diff
@@ -37,12 +37,19 @@ service Metadata {
...
     rpc CreateModel(Model) returns (Empty);
     rpc GetModels(stream Name) returns (stream Model);
+    rpc SetResourceStatus(StatusRequest) returns (Empty);
 }
 
 message Name {
     string name = 1;
 }
 
+message StatusRequest {
```
Comment (Sami1309) [inline ff/serving/metadata/proto/metadata.proto:47]: setStatusRequest

File: ff/serving/metadata/proto/metadata.proto [hunk 19]
```diff
@@ -37,12 +37,19 @@ service Metadata {
...
+    NameVariant resource = 1;
+    string type = 2;
```
Comment (Sami1309) [inline ff/serving/metadata/proto/metadata.proto:49]: ambiguous? (resourceType)

**Review by simba-git — COMMENTED** [review]
Just test types in online table and its good to go

**Review by simba-git — APPROVED** [review]
